configuration/config2/EnvironmentNode.cpp

```diff
@@ -0,0 +1,385 @@
+    }
+
+    rc = m_pNodeValue->setValue(value, &status, force);
+    return rc;
+}
+
+
+void EnvironmentNode::validate(Status &status, bool includeChildren) const
+{
+    //
+    // Check node value
+    if (m_pNodeValue)
+    {
+        m_pNodeValue->validate(status, "");
+    }
+
+    //
+    // Check any attributes
+    for (auto attrIt = m_attributes.begin(); attrIt != m_attributes.end(); ++attrIt)
+    {
+        attrIt->second->validate(status, m_id);
+
+        //
+        // If this value must be unique, make sure it is
+        if (attrIt->second->getSchemaValue()->isUniqueValue())
+        {
+            bool found = false;
+            std::vector<std::string> allValues = attrIt->second->getAllValues();
+            std::set<std::string> unquieValues;
+            for (auto it = allValues.begin(); it != allValues.end() && !found; ++it)
+            {
+                auto ret = unquieValues.insert(*it);
+                found = ret.second;
+            }
+
+            if (found)
+            {
+                status.addUniqueMsg(statusMsg::error, m_id, attrIt->second->getName(), "", "Attribute value must be unique");
+            }
+        }
+
+        //
+        // Does this value need to be from another set of values?
+        if (attrIt->second->getSchemaValue()->isFromUniqueValueSet())
+        {
+            bool found = false;
+            std::vector<std::string> allValues = attrIt->second->getSchemaValue()->getAllKeyRefValues();
+            for (auto it = allValues.begin(); it != allValues.end() && !found; ++it)
+                found = *it == attrIt->second->getValue();
+            if (!found)
+            {
+                status.addMsg(statusMsg::error, m_id, attrIt->second->getName(), "", "Attribute value must be from a unique set");
+            }
+        }
+    }
+
+    //
+    // Now check all children
+    if (includeChildren)
+    {
+        for (auto childIt = m_children.begin(); childIt != m_children.end(); ++childIt)
+        {
+            childIt->second->validate(status);
+        }
+    }
+}
+
+
+std::vector<std::string> EnvironmentNode::getAllAttributeValues(const std::string &attrName) const
+{
+    std::vector<std::string> values;
+    std::shared_ptr<EnvironmentNode> pParentNode = m_pParent.lock();
+    if (pParentNode)
+    {
+        std::vector<std::shared_ptr<EnvironmentNode>> nodes = pParentNode->getChildren(m_name);
+        for (auto it = nodes.begin(); it != nodes.end(); ++it)
+        {
+            values.push_back((*it)->getAttributeValue(attrName));
+        }
+    }
+    return values;
+}
+
+
+const std::shared_ptr<EnvironmentValue> EnvironmentNode::getAttribute(const std::string &name) const
+{
+    std::shared_ptr<EnvironmentValue> pValue;
+    auto it = m_attributes.find(name);
+    if (it != m_attributes.end())
+    {
+        pValue = it->second;
+    }
+    return pValue;
+}
+
+
+std::vector<std::shared_ptr<SchemaItem>> EnvironmentNode::getInsertableItems() const
```

Comment:
Don't quite understand what "insertable" means.

Reply:
Insertable items is a collection of components or other environment items that can be added to the current node. For example, an Instance for an additional node for an ESP service. It could also be an ESP service as well. The method takes into account the number of allowed instances specified in the XSD file.